### ISSUE TITLE
Support for running a script on workspace delete

### ISSUE THREAD

**Issue by evilhamsterman** [coder/coder#13387]
Right now there's support to run scripts on workspace startup and stop, but it would also be nice to run a script when a workspace is deleted. 

For example we have workspaces that are persistent VMs, when they are started they install an register an XDR agent. We'd like to run a script to uninstall the agent which also unregisters it when they are deleted. We can run a script to do so when the workspace is stopped, but then we have to reinstall it on start and uninstall it on stop, and that creates a lot of unnecessary churn.

**Comment by kylecarbs** [comment]
You can do this via Terraform.

Use a `null_resource` and check if `data.coder_workspace.me.transition === "delete"`, then run a script.

**Comment by evilhamsterman** [comment]
Can you run the script in the workspace like the `run_on_stop` option for `coder_script`?

**Comment by matifali** [comment]
I guess to run a script on delete transition we need an attribute `run_on_delete` for `coder_script`. It will only work if the workspace is in running state. 

@mafredri could have some thoughts here.

**Comment by mafredri** [comment]
@matifali Yes, `run_on_delete` and a running workspace are prerequisites. Ideally we would also have #6175 to ensure proper execution of the script.

Having the `run_on_delete` property gives a nice way for us to surface a warning if a stopped workspace is deleted and there is a `run_on_delete` script. Then we can suggest starting the workspace first. Whether or not this is a nice UX flow is another question. I can imagine users first stopping their workspaces before issuing delete, only to be asked to start it again.

Alternatively, we could support this via `run_on_stop` by setting an environment variable `CODER_WORKSPACE_STOP_REASON=stop|delete` and letting the template author handle it.

**Comment by matifali** [comment]
This could also be used to backup workspace contents for compliance/security purposed.

**Comment by wSedlacek** [comment]
We use coder to create namespaces with permissions for devs.
Devs can add resources to that namespace while they are running.
But since there is no hook when the workspace is deleted and the namespace should be deleted we end up trying to delete the namespace but failing due to finalizers being blocked when not tearing down resources in a specific order.

tl;dr A on delete solution would really help my team out. Perhaps being able to spin up a 1 off job so the deployment doesn't have to be running to complete it.

**Comment by matifali** [comment]
> tl;dr A on delete solution would really help my team out. Perhaps being able to spin up a 1 off job so the deployment doesn't have to be running to complete it.

@wSedlacek 
You should be able to run a script using `local_provisioner` and a `null_resource` and check if `data.coder_workspace.me.transition === "delete"`.
